src/navigation/navigation-events.js

```diff
@@ -55,7 +55,11 @@ export function navigateToUrl(obj) {
     destination.pathname === current.pathname &&
     destination.search === current.search
   ) {
-    window.location.hash = destination.hash;
+    // fix Firefox automatically add hash sign in the end URL
+    // if setting window.location.hash to empty string or #
+    if (destination.hash && destination.hash !== "#") {
```

Comment:
This if statement makes the `navigateToUrl` function a no-op, which I need to think about a bit. I think the change is fine. This function is a really important one, and so I'm going to suggest a nitpicky change to avoid any javascript truthiness weirdness. I believe the `.hash` property is always a string, but to be safe we'd test this in IE and other browsers

```suggestion
    if (destination.hash.length > 0 && destination.hash !== "#") {
```